src/main/java/com/thealgorithms/strings/AhoCorasick.java

```diff
@@ -0,0 +1,236 @@
+            /*
+             * res is the list of list containing the indexes of words in patterns/dictionary
+             * list index represents word.
+             * eg - list[0] contains the list of start-index of word pattern[0]
+             *      list[1] contains the list of start-index of word pattern[1]
+             *      ......
+             *      ......
+             *      list[n] contains the list of start-index of word pattern[n]
+             */
+            ArrayList<ArrayList<Integer>> res = new ArrayList<>(patterns.length); // Stores positions where patterns are found in the text
+            Node parent = root; // Start searching from the root node
+
+            for (int i = 0; i < patterns.length; i++) {
+                res.add(new ArrayList<Integer>()); // Initialize a list to store positions of the current pattern
+            }
+
+            for (int i = 0; i < text.length(); i++) {
+                char ch = text.charAt(i); // Get the current character in the text
+
+                // Check if the current node has a child for the current character
+                if (parent.getChild().containsKey(ch)) {
+                    parent = parent.getChild().get(ch); // Update the current node to the child node
+
+                    // If the current node represents a pattern, record its position in res
+                    if (parent.getPatternInd() > -1) {
+                        res.get(parent.getPatternInd()).add(i);
+                    }
+
+                    Node outputLink = parent.getOutputLink();
+                    // Follow output links to find and record positions of other patterns
+                    while (outputLink != null) {
+                        res.get(outputLink.getPatternInd()).add(i);
+                        outputLink = outputLink.getOutputLink();
+                    }
+                } else {
+                    // If no child node exists for the character, backtrack using suffix links
+                    while (parent != root && !parent.getChild().containsKey(ch)) {
+                        parent = parent.getSuffixLink();
+                    }
+                    if (parent.getChild().containsKey(ch)) {
+                        i--; // Decrement i to reprocess the same character
+                    }
+                }
+            }
+            setUpStartPoints(res);
+            return res;
+        }
+        // by default res contains end-points. This function converts those
+        // endpoints to start points
+        private void setUpStartPoints(ArrayList<ArrayList<Integer>> res) {
+            for (int i = 0; i < patterns.length; i++) {
+                if (!res.get(i).isEmpty()) {
+                    for (int j = 0; j < res.get(i).size(); j++) {
+                        int endpoint = res.get(i).get(j);
+                        res.get(i).set(j, endpoint - patterns[i].length() + 1);
+                    }
+                }

```

Comment:
I think this `if` statement is not needed here. If it would be `false`, then the loop has length 0 and nothing happens anyway.
```suggestion
                for (int j = 0; j < res.get(i).size(); j++) {
                    int endpoint = res.get(i).get(j);
                    res.get(i).set(j, endpoint - patterns[i].length() + 1);
                }
```

Reply:
done.